### PR TITLE
Validations

### PR DESCRIPTION
**Pull request by lu-jim**
## Purpose:
- Use validations for models.
- Write unit tests for models.
- Secure app from n+1 problems.

## Features:
- Added validations for user and post
- Created tests for these models
- Validated the app with Bullet to avoid N+1 errors

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -1,5 +1,33 @@
...
+      post_counter: 0
+    )
+  end
+
+  before { subject.save }
+
+  it 'is valid with valid attributes' do
+    expect(subject).to be_valid
+  end
+
```
Comment (SajjadAhmad14) [inline spec/models/user_spec.rb:31]: Please add unit test for `#recent_posts` of the User model.

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -1,5 +1,52 @@
...
+  it 'is not valid with likes_counter negative' do
+    subject.like_counter = -1
+    expect(subject).to_not be_valid
+  end
+
+  it 'is not valid with likes_counter not integer' do
+    subject.like_counter = 1.618
+    expect(subject).to_not be_valid
+  end
+
```
Comment (SajjadAhmad14) [inline spec/models/post_spec.rb:63]: Please add unit tests for `#recent_comments` and `#update_post_counter` of Post model.

File: app/controllers/posts_controller.rb [hunk 3]
```diff
@@ -17,14 +17,16 @@ def show
   def create
```
Comment (SajjadAhmad14) [inline app/controllers/posts_controller.rb:17]: Please solve the `N+1` problem when fetching all posts and their comments for a user in the `#index` action `PostsController`. You may learn more about the `N+1` problem [here](https://medium.com/idopterlabs/solving-the-n-1-problem-in-rails-b5ea677e78e0).

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
## Required Changes ♻️
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hi Luis, 
 ### You have been doing a tremendous job so far :ok_hand:

## To Highlight 🎉
Git flow is followed. :heavy_check_mark:
README is professional. :heavy_check_mark:
## To Improve 🙏

- [x] Please check inline comments under the review

## Optional suggestions
_Comments with the [OPTIONAL] prefix are not a must to implement however, It's strongly recommended to work on them as they can make your code better._

## Happy coding!👏👏👏

#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
#### Please remember to tag me in your question so I can receive the notification.
**_Note: Please do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by bimbolabuari — APPROVED** [review]
# Project Approved 🚀

Hi @lu-jim 🥇,

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**